---
title: "TFutils motivations: overview of bioinformatics of transcription factors and their binding sites"
author: "Shweta G., et al."
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{TF utils motivations}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
bibliography: TFutils.bib
---

```{r setup,echo=FALSE,results="hide"}
suppressPackageStartupMessages({
suppressMessages({
library(TFutils)
library(gwascat)
})
})
```

# Introduction

TFutils addresses very general problems of annotation and location
of TFs and their targets.

# TF involvement in GWAS findings

The following illustrates aspects of combinatorics of TFs in
conjunction with GWAS findings for a collection of diseases (@Maurano2012).

![](images/mauranoTFcombs.png)

## Looking for genes association with POU2F1 in GO

Let's consider how to identify the genes in the POU2F1 pathway and check 
for GWAS hits.  According to GeneCards, the POU2F1 TF is involved with the
glucocorticoid receptor network and the ATF-2 TF network.

Let's try GO.

```{r lkgo}
library(GO.db)
allt = keys(GO.db, keytype="TERM")
grep("glucocorticoid", allt, value=TRUE)
select(GO.db, keys="glucocorticoid receptor signaling pathway", 
  columns=c("GOID", "SYMBOL"), keytype="TERM")
```

Then check in ebicat

```{r dogwas}
library(gwascat)
data(ebicat37)
ebicat37
```

## Using MSigDb

```{r dogse}
library(GSEABase)
getGmt(system.file("msigdb_extracts/POU1F1_Q6.gmt", package="TFutils")) -> poum
poum[[1]] -> p1
library(org.Hs.eg.db)
geneIdType(p1) = SymbolIdentifier()
p2 = p1
geneIdType(p2) = EntrezIdentifier("org.Hs.eg.db")
```
search the gwascatalog for hits in these genes

Let's say that MsigDb is our preferred map from TFs to gene sets (as TF targets)

Should we make a database that links from familiar symbols like VDR to targets of VDR
or from ORMDL3 to any TF that happens to bind near it....

MSigDb has a RESTful interfaces?  tft gmt has been acquired

canonical HGNC names for TFs derived from tfname (remove suffix for example) and bridge
to the M numbers used in FIMO

compare the FIMO binding profile for a TF to the genes called 'targets'

binding profiles are in bed files -> granges on some chromosome
also:  genes(TxDb.Hsapiens.UCSC.hg19.knownGene) -> a granges ... do the places of high binding
coincide with the promoter regions or any regions of the genes called 'targets'

```{r dofimo}
library(TFutils)
data("named_tf")
fimo_bs = getBS("VDR", named_tf, GenomicRanges::GRanges("chr1", IRanges::IRanges(1,25000)),importFIMO_local)
```

```{do txdb}
library(feather)
library(dplyr)
library(magrittr)
library(data.table)
tfhash = read_feather(system.file("feather/tfhash2.feather", package = "TFutils"))
tfhash_vdr = tfhash %>% filter(tfname %like% 'VDR')
#Grab some target genes of VDR
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txdb
seqlevels(txdb) <- "chr1" #looking at chromosome 1 
keys = as.character(tfhash_vdr$targets) # these are geneids
select(txdb, keys = keys, columns="TXNAME", keytype="GENEID")
GR <- transcripts(txdb, filter=list(tx_chrom = "chr1"))
library(GenomicRanges)
subsetByOverlaps(fimo_bs, GR)
cols <- c("TXNAME", "TXSTRAND", "TXCHROM")
res = select(txdb, keys=keys, columns=cols, keytype="GENEID")
GR <- transcripts(txdb, filter=list(tx_chrom = "chr1", tx_name=res))
```



# References
